contracts/src/PDPService.sol

```diff
@@ -231,9 +229,24 @@ contract PDPService {
         candidate = acc + sumTreeCounts[setId][searchPtr];
         if (candidate <= leafIndex) {
             // Choose right 
-            return (searchPtr + 1, leafIndex - candidate); 
+            return RootIdAndOffset(searchPtr + 1, leafIndex - candidate); 
         } // Choose left
-        return (searchPtr, leafIndex - acc);
+        return RootIdAndOffset(searchPtr, leafIndex - acc);
+    }
+
+    // findRoot returns the root id for a given leaf index and the leaf's offset within
+    // the rootId.
+    // 
+    // It does this by running a binary search over the logical array
+    // To do this efficiently we walk the sumtree.
+    function findRootIds(uint256 setId, uint256[] calldata leafIndexs) public view returns (RootIdAndOffset[] memory) { 
+        // The top of the sumtree is the largest power of 2 less than the number of roots
+        uint256 top = 256 - BitOps.clz(nextRootId[setId]);
```

Comment:
Maybe this is premature optimization -- the only reason this is here and not in findOneRootId is for the potential gas savings